Adds hyperscale image for pg14

gurkanindibay opened the pull request:
Created the image for pg 14. I will need to create for each pg version. 

I haven't added a user with the privileges you sent. I will add it after you first pre-review
you can create a new docker image in the docker repo with the command below
**PG14**
docker build -t hyperscale_image_14 -f hyperscale/pg14/Dockerfile .

**PG13**
docker build -t hyperscale_image_13 -f hyperscale/pg14/Dockerfile .

**PG12**
docker build -t hyperscale_image_12 -f hyperscale/pg12/Dockerfile .

and run with the following command (replacing with the pg version you give above)

docker run --name test1 -p 5432:5432 -e CITUS_PASSWORD="Citus123" hyperscale_image_<pg-version>

I compared the plugins between the docker image I created and th list you've sent and the comparison excel is below
[Hyperscale Image Extensions.xlsx](https://github.com/citusdata/docker/files/7487247/Hyperscale.Image.Extensions.xlsx)

marcocitus commented on 003-other-extensions.sql:
```diff
@@ -0,0 +1,11 @@
+create EXTENSION pg_partman;
+create EXTENSION tdigest;
+create EXTENSION mysql_fdw;
+-- create EXTENSION pg_cron;
```
we should create pg_cron as well

gurkanindibay replied:
Fixed

gurkanindibay commented on 003-other-extensions.sql:
```diff
@@ -0,0 +1,11 @@
+create EXTENSION hstore_plperl;
+create EXTENSION jsonb_plperl;
+create EXTENSION plperlu;
+create EXTENSION jsonb_plperlu;
```
I can not remove plperl since it is required by hstore_plperl

2021-11-08 06:29:59.740 UTC [65] ERROR:  required extension "plperl" is not installed
2021-11-08 06:29:59.740 UTC [65] HINT:  Use CREATE EXTENSION ... CASCADE to install required extensions too.
2021-11-08 06:29:59.740 UTC [65] STATEMENT:  create EXTENSION hstore_plperl;
psql:/docker-entrypoint-initdb.d/003-other-extensions.sql:3: ERROR:  required extension "plperl" is not installed
HINT:  Use CREATE EXTENSION ... CASCADE to install required extensions too.

gurkanindibay replied:
Added your suggested extensions

marcocitus replied:
You can remove hstore_plperl too. It's installed on Hyperscale, but there is no way for the user to create it and it is not pre-created.

gurkanindibay replied:
Removed plperl hstore_plperl and jsonb_plperl

gurkanindibay commented on 002-create-postgis-extension.sql:
```diff
@@ -0,0 +1,19 @@
+-- example rule data set
+CREATE EXTENSION address_standardizer_data_us;
+-- Enable US Tiger Geocoder
+CREATE EXTENSION postgis_tiger_geocoder;
```
I didn't understand this comment. Should I add this function and call it? This is like a white/blacklist function I think

marcocitus replied:
> . Should I add this function and call it? 

Only add the function such that users can call it.

gurkanindibay replied:
Added functions

gurkanindibay commented on 003-other-extensions.sql:
```diff
@@ -0,0 +1,11 @@
+create EXTENSION pg_freespacemap;
+create EXTENSION pg_partman;
+create EXTENSION tdigest;
+create EXTENSION mysql_fdw;
```
Removed it

gurkanindibay commented on hyperscale/Dockerfile:
```diff
@@ -0,0 +1,58 @@
+RUN echo "shared_preload_libraries='citus,pg_cron,pg_stat_statements'" >> /usr/share/postgresql/postgresql.conf.sample
+
+# add pg_cron into default PostgreSQL config
+RUN echo "cron.database_name = 'postgres'" >> /usr/share/postgresql/postgresql.conf.sample
```
I created a citus database but I couldn't make pg_cron work with the citus database. I got the below error
```
2021-11-08 10:35:59.191 UTC [72] STATEMENT:  CREATE EXTENSION pg_cron;
psql:/docker-entrypoint-initdb.d/005-other-extensions.sql:9: ERROR:  can only create extension in database citus
DETAIL:  Jobs must be scheduled from the database configured in cron.database_name, since the pg_cron background worker reads job description
s from this database.
HINT:  Add cron.database_name = 'postgres' in postgresql.conf to use the current database.
CONTEXT:  PL/pgSQL function inline_code_block line 4 at RAISE
```

marcocitus replied:
That probably means we are not creating the extensions in the citus database. 

Is there a way we could execute the SQL files below in the `citus` database?

marcocitus replied:
It seems we should be able to set the POSTGRES_DB environment variable
https://github.com/docker-library/postgres/blob/dd84d6fccbab39bb67fb8f80b9fb2bfc5cb0c992/docker-entrypoint.sh#L191

gurkanindibay replied:
Fixed

gurkanindibay commented on hyperscale/006-grants.sql:
```diff
@@ -0,0 +1,32 @@
+GRANT EXECUTE ON ALL FUNCTIONS IN SCHEMA partman TO citus WITH GRANT OPTION;
+GRANT EXECUTE ON ALL PROCEDURES IN SCHEMA partman TO citus WITH GRANT OPTION;
+GRANT USAGE ON FOREIGN DATA WRAPPER dblink_fdw TO citus WITH GRANT OPTION;
+-- TODO Having problems installing below functions
+-- GRANT EXECUTE ON FUNCTION public.pg_replication_origin_create(text) TO citus;
+-- GRANT EXECUTE ON FUNCTION public.pg_replication_origin_session_progress(bool) TO citus;
+-- GRANT EXECUTE ON FUNCTION public.pg_replication_origin_progress(text, bool) TO citus;
+-- GRANT EXECUTE ON FUNCTION public.pg_replication_origin_drop(text) TO citus;
+-- GRANT EXECUTE ON FUNCTION public.pg_replication_origin_xact_setup(pg_lsn, timestamptz) TO citus;
+-- GRANT EXECUTE ON FUNCTION public.pg_replication_origin_session_setup(text) TO citus;
```
I 'm getting errors while executing these functions. I'm trying to find function scripts to install

gurkanindibay replied:
Added functions and enabled grants

marcocitus commented on hyperscale/006-other_extensions.sql:
```diff
@@ -0,0 +1,25 @@
+CREATE EXTENSION hstore;
+CREATE EXTENSION plperlu;
+CREATE EXTENSION jsonb_plperlu;
```
can remove plperlu and jsonb_plperlu, they are not created on Hyperscale

gurkanindibay replied:
Removed

gurkanindibay replied:
You're right I fixed them

marcocitus commented on hyperscale/008-grant_usage_fdw.sql:
```diff
@@ -0,0 +1,2 @@
+-- source https://msdata.visualstudio.com/Database%20Systems/_git/CitusMigration?path=/instance/lib/grant_usage_fdw.sql
```
maybe don't include links to the internal repo in this public repo

gurkanindibay replied:
Removed links

marcocitus commented on hyperscale/002-db-user-creation.sql:
```diff
@@ -0,0 +1,2 @@
+create user citus with encrypted password 'Citus123';
```
is it possible to configure the password via a setting somehow?

gurkanindibay replied:
Parameterized password. Now we can set the password in docker run

docker run --name image_name -p 5432:5432 -e POSTGRES_PASSWORD="Citus124" -e CITUS_PASSWORD="1234"  hyperscale_image

hanefi commented on hyperscale/Dockerfile:
```diff
@@ -0,0 +1,70 @@
+ENV CITUS_VERSION ${VERSION}.citus-1
+ENV POSTGRES_DB citus
+
+# install Citus
+RUN apt-get update \
+    && apt-get install -y --no-install-recommends \
+       ca-certificates \
+       curl \
+    && curl -s https://install.citusdata.com/community/deb.sh | bash \
+    && apt-get install -y postgresql-$PG_MAJOR-citus-10.2=$CITUS_VERSION \
+                          postgresql-$PG_MAJOR-hll=2.16.citus-1 \
+                          postgresql-$PG_MAJOR-topn=2.4.0 \
+    && apt-get purge -y --auto-remove curl \
+    && rm -rf /var/lib/apt/lists/*
+
+# install other extensions
+RUN     apt-get update \
+      &&apt-get install -y postgresql-14-postgis-3 \
+            postgresql-plperl-14 \
+            postgresql-contrib \
+            postgresql-14-partman \
+            postgresql-14-tdigest \
+            postgresql-14-mysql-fdw
+
+RUN apt-get -y install postgresql-14-cron
```
I suggest that you combine all the package installation steps in here.

gurkanindibay replied:
Combined with the upper command

hanefi commented on hyperscale/Dockerfile:
```diff
@@ -0,0 +1,70 @@
+    && rm -rf /var/lib/apt/lists/*
+
+# install other extensions
+RUN     apt-get update \
+      &&apt-get install -y postgresql-14-postgis-3 \
+            postgresql-plperl-14 \
+            postgresql-contrib \
+            postgresql-14-partman \
+            postgresql-14-tdigest \
+            postgresql-14-mysql-fdw
```
```suggestion
RUN apt-get update \
    && apt-get install -y postgresql-14-postgis-3 \
                          postgresql-plperl-14 \
                          postgresql-contrib \
                          postgresql-14-partman \
                          postgresql-14-tdigest \
                          postgresql-14-mysql-fdw
```

gurkanindibay replied:
Fixed

hanefi commented on hyperscale/Dockerfile:
```diff
@@ -0,0 +1,70 @@
+COPY hyperscale/007-pg_cron_pg_partman_setup.sql /docker-entrypoint-initdb.d/
+COPY hyperscale/008-grant_usage_fdw.sql /docker-entrypoint-initdb.d/
+COPY hyperscale/009-other_grants.sql /docker-entrypoint-initdb.d/
+COPY hyperscale/010-stat_statements_reset.sql /docker-entrypoint-initdb.d/
+
+
+
+
+
+# add health check script
```
remove extra empty lines

```suggestion
COPY hyperscale/010-stat_statements_reset.sql /docker-entrypoint-initdb.d/

# add health check script
```

hanefi commented on hyperscale/001-db-user-creation.sql:
```diff
@@ -0,0 +1,3 @@
+create user citus with encrypted password 'Citus123';
```
```suggestion
CREATE USER citus WITH ENCRYPTED PASSWORD 'Citus123';
```

It is better to keep capital letters when using sql keywords. Please go over other `.sql` files for similar styles.

gurkanindibay replied:
Fixed

gurkanindibay commented on hyperscale/005-create_postgis_extension.sql:
```diff
@@ -0,0 +1,21 @@
+-- example rule data set
+CREATE EXTENSION address_standardizer_data_us;
+-- Enable US Tiger Geocoder
+CREATE EXTENSION postgis_tiger_geocoder;
```
Uncommented all lines and added a comment

hanefi commented on hyperscale/002-create_extension_functions.sql:
```diff
@@ -0,0 +1,80 @@
+'tsm_system_time') THEN raise 'not allowed to drop this extension';
+END IF;
+EXECUTE pg_catalog.format('DROP EXTENSION %I;', extname);
+END; $function$;
```
```suggestion
END; $function$;

```

hanefi commented on hyperscale/004-postgres_replication_origins.sql:
```diff
@@ -0,0 +1,49 @@
+GRANT EXECUTE ON FUNCTION public.pg_replication_origin_progress(text, bool) TO citus;
+GRANT EXECUTE ON FUNCTION public.pg_replication_origin_drop(text) TO citus;
+GRANT EXECUTE ON FUNCTION public.pg_replication_origin_xact_setup(pg_lsn, timestamptz) TO citus;
+GRANT EXECUTE ON FUNCTION public.pg_replication_origin_session_setup(text) TO citus;
```
```suggestion
GRANT EXECUTE ON FUNCTION public.pg_replication_origin_session_setup(text) TO citus;

```

hanefi commented on hyperscale/005-create_postgis_extension.sql:
```diff
@@ -0,0 +1,26 @@
+----------------------------------------------------------------------------------
```
Why do we need this file if we are sure that all the contents will be commented out?

gurkanindibay replied:
I keep it in case developers may want to use this file to enable it

marcocitus replied:
I think they can easily add additional lines to hyperscale/006-other_extensions.sql, I'd prefer removing this one.

hanefi replied:
Let's remove this file. We may even need to renumber all the files `006-other_extensions.sql` to `010-stat_statements_reset.sql` after removing this one.

gurkanindibay replied:
Removed

hanefi commented on hyperscale/007-pg_cron_pg_partman_setup.sql:
```diff
@@ -0,0 +1,9 @@
+GRANT ALL ON SCHEMA partman TO citus WITH GRANT OPTION;
+GRANT ALL ON ALL TABLES IN SCHEMA partman TO citus WITH GRANT OPTION;
+GRANT EXECUTE ON ALL FUNCTIONS IN SCHEMA partman TO citus WITH GRANT OPTION;
+GRANT EXECUTE ON ALL PROCEDURES IN SCHEMA partman TO citus WITH GRANT OPTION;
```
```suggestion
GRANT EXECUTE ON ALL PROCEDURES IN SCHEMA partman TO citus WITH GRANT OPTION;

```

hanefi commented on hyperscale/008-grant_usage_fdw.sql:
```diff
@@ -0,0 +1 @@
+GRANT USAGE ON FOREIGN DATA WRAPPER dblink_fdw TO citus WITH GRANT OPTION;
```
```suggestion
GRANT USAGE ON FOREIGN DATA WRAPPER dblink_fdw TO citus WITH GRANT OPTION;

```

hanefi commented on hyperscale/009-other_grants.sql:
```diff
@@ -0,0 +1,18 @@
+GRANT EXECUTE ON FUNCTION public.pg_replication_origin_progress(text, bool) TO citus;
+GRANT EXECUTE ON FUNCTION public.pg_replication_origin_drop(text) TO citus;
+GRANT EXECUTE ON FUNCTION public.pg_replication_origin_xact_setup(pg_lsn, timestamptz) TO citus;
+GRANT EXECUTE ON FUNCTION public.pg_replication_origin_session_setup(text) TO citus;
```
```suggestion
GRANT EXECUTE ON FUNCTION public.pg_replication_origin_session_setup(text) TO citus;

```

hanefi commented on hyperscale/010-stat_statements_reset.sql:
```diff
@@ -0,0 +1,2 @@
+GRANT EXECUTE ON FUNCTION pg_stat_statements_reset TO citus WITH GRANT OPTION;
+GRANT EXECUTE ON FUNCTION citus_stat_statements_reset TO citus WITH GRANT OPTION;
```
```suggestion
GRANT EXECUTE ON FUNCTION citus_stat_statements_reset TO citus WITH GRANT OPTION;

```

hanefi commented on hyperscale/Dockerfile:
```diff
@@ -0,0 +1,76 @@
+ENV CITUS_VERSION ${VERSION}.citus-1
+ENV POSTGRES_DB citus
+
+# install Citus
+RUN apt-get update \
+    && apt-get install -y --no-install-recommends \
+       ca-certificates \
+       curl \
+    && curl -s https://install.citusdata.com/community/deb.sh | bash \
+    && apt-get install -y postgresql-$PG_MAJOR-citus-10.2=$CITUS_VERSION \
+                          postgresql-$PG_MAJOR-hll=2.16.citus-1 \
+                          postgresql-$PG_MAJOR-topn=2.4.0 \
+    && apt-get purge -y --auto-remove curl \
+    && rm -rf /var/lib/apt/lists/*
+
+# install other extensions
+RUN apt-get update \
+    &&apt-get install -y postgresql-14-postgis-3 \
+                         postgresql-plperl-14 \
+                         postgresql-contrib \
+                         postgresql-14-partman \
+                         postgresql-14-tdigest \
+                         postgresql-14-mysql-fdw \
+                         postgresql-14-cron
```
Please combine these `apt-get` `RUN` steps

gurkanindibay replied:
Fixed

hanefi commented on hyperscale/Dockerfile:
```diff
@@ -0,0 +1,76 @@
+                         postgresql-14-mysql-fdw \
+                         postgresql-14-cron
+
+# add citus to default PostgreSQL config
```
This comment is slightly misleading, and wrong.

```suggestion
# set shared_preload_libraries to the list of extensions that needs to be loaded when postgresql starts
```

hanefi commented on hyperscale/Dockerfile:
```diff
@@ -0,0 +1,76 @@
+# add citus to default PostgreSQL config
+RUN echo "shared_preload_libraries='citus,pg_cron,pg_stat_statements'" >> /usr/share/postgresql/postgresql.conf.sample
+
+# add pg_cron into default PostgreSQL config
```
```suggestion
# set default database name for pg_cron extension
```

hanefi commented on hyperscale/Dockerfile:
```diff
@@ -0,0 +1,76 @@
+COPY hyperscale/007-pg_cron_pg_partman_setup.sql /docker-entrypoint-initdb.d/
+COPY hyperscale/008-grant_usage_fdw.sql /docker-entrypoint-initdb.d/
+COPY hyperscale/009-other_grants.sql /docker-entrypoint-initdb.d/
+COPY hyperscale/010-stat_statements_reset.sql /docker-entrypoint-initdb.d/
+
+
+# add health check script
+COPY pg_healthcheck wait-for-manager.sh /
+
+
+RUN chmod +x /wait-for-manager.sh
```
```suggestion
COPY hyperscale/010-stat_statements_reset.sql /docker-entrypoint-initdb.d/

# add health check script
COPY pg_healthcheck wait-for-manager.sh /

RUN chmod +x /wait-for-manager.sh
```

remove extra spaces

hanefi commented on hyperscale/Dockerfile:
```diff
@@ -0,0 +1,73 @@
+                          postgresql-14-mysql-fdw \
+                          postgresql-14-cron \
+    && apt-get purge -y --auto-remove curl \
+    && rm -rf /var/lib/apt/lists/*
+
+
+
+# set shared_preload_libraries to the list of extensions that needs to be loaded when postgresql starts
```
```suggestion
    && rm -rf /var/lib/apt/lists/*

# set shared_preload_libraries to the list of extensions that needs to be loaded when postgresql starts
```

remove extra spaces

hanefi commented on hyperscale/Dockerfile:
```diff
@@ -0,0 +1,73 @@
+
+# This block is required to be able to define the password for user citus. With the below script citus password can be
+# set using 'CITUS_PASSWORD' env variable set in docker run phase
+RUN apt-get update && apt-get install -y  sudo && adduser postgres sudo
```
You need to run `rm -rf /var/lib/apt/lists/*` after using `apt-get` cli to be able to shrink image size.

```suggestion
RUN apt-get update 
    && apt-get install -y sudo
    && adduser postgres sudo
    && rm -rf /var/lib/apt/lists/*
```

However, I do not really understand why you want postgres to have sudo privileges. It is not really recommended to do this.

gurkanindibay replied:
@marcocitus wanted me to parametize the password change. I need to change the password in docker run phase to be able to replace password that is the sole solution that I could find.

hanefi replied:
An excerpt from [Docker docs](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#user)

> Avoid installing or using sudo as it has unpredictable TTY and signal-forwarding behavior that can cause problems. If you absolutely need functionality similar to sudo, such as initializing the daemon as root but running it as non-root, consider using “gosu”.

Please do not install sudo without making sure we are ok with all the consequences that comes with it.

I am 100% sure we do not give sudo permissions to postgres user on HyperScale, and we should mimic the behaviour here as well.

gurkanindibay replied:
Removed sudo and changed ownership folder for migration folder for  postgres user

hanefi commented on hyperscale/Dockerfile:
```diff
@@ -0,0 +1,73 @@
+# This block is required to be able to define the password for user citus. With the below script citus password can be
+# set using 'CITUS_PASSWORD' env variable set in docker run phase
+RUN apt-get update && apt-get install -y  sudo && adduser postgres sudo
+RUN echo 'postgres     ALL=(ALL) NOPASSWD:ALL' >> '/etc/sudoers'
```
I do not think that it is wise to add postgres user as a sudoer. Are you sure this is safe to do?

gurkanindibay replied:
I think this is for development purposes not for production. Therefore, there is no harm for postgres user to have sudo permission. Since it is solely for development purposes, it may make our life easier for the further developments. 
@marcocitus what do you think?

marcocitus commented on hyperscale/009-other_grants.sql:
```diff
@@ -0,0 +1,18 @@
+GRANT EXECUTE ON FUNCTION public.pg_replication_origin_progress(text, bool) TO citus;
+GRANT EXECUTE ON FUNCTION public.pg_replication_origin_drop(text) TO citus;
+GRANT EXECUTE ON FUNCTION public.pg_replication_origin_xact_setup(pg_lsn, timestamptz) TO citus;
+GRANT EXECUTE ON FUNCTION public.pg_replication_origin_session_setup(text) TO citus;
```
The last 6 grants are already in hyperscale/004-postgres_replication_origins.sql

gurkanindibay replied:
Removed

marcocitus commented on hyperscale/Dockerfile:
```diff
@@ -0,0 +1,73 @@
+# set using 'CITUS_PASSWORD' env variable set in docker run phase
+RUN apt-get update && apt-get install -y  sudo && adduser postgres sudo
+RUN echo 'postgres     ALL=(ALL) NOPASSWD:ALL' >> '/etc/sudoers'
+ARG CITUS_PWD_COMMAND='\\tsudo sed "s\/###CITUS_PASSWORD###\/${CITUS_PASSWORD}\/g" -i \/docker-entrypoint-initdb.d\/001-db-user-creation.sql\n'
```
What do you think of (re)using ${POSTGRES_PASSWORD} here?

That way users only have to specify one password when they use docker run.

gurkanindibay replied:
Ok I can change this way

gurkanindibay replied:
Changed as you suggested

marcocitus commented on hyperscale/001-db-user-creation.sql:
```diff
@@ -0,0 +1,3 @@
+CREATE USER citus WITH ENCRYPTED PASSWORD '###CITUS_PASSWORD###';
```
thanks for figuring this out!

gurkanindibay replied:
@marcocitus  do you want me to remove 005-create_postgis_extension.sql or delete the content of the file?

hanefi commented on hyperscale/Dockerfile:
```diff
@@ -0,0 +1,68 @@
+COPY hyperscale/009-stat_statements_reset.sql /docker-entrypoint-initdb.d/
+
+# add health check script
+COPY pg_healthcheck wait-for-manager.sh /
+
+RUN chmod +x /wait-for-manager.sh
```
Why do we need `+x` for `wait-for-manager.sh` but not `pg_healthcheck` ?

This seems a bit weird.

gurkanindibay replied:
I haven't find any reason since waitformanager already has 'x' permission as you see below so removed it

![image](https://user-images.githubusercontent.com/1596331/141444870-56b6fbae-eabf-41ff-9e4c-d6d0427d83c6.png)

marcocitus reviewed (commented):
This is great!

I wonder whether there are any settings we should replicate, but from a quick look I did not see anything super important.

gurkanindibay commented on hyperscale/005-other_extensions.sql:
```diff
@@ -0,0 +1,23 @@
+CREATE EXTENSION btree_gist;
+CREATE EXTENSION citext;
+CREATE EXTENSION dblink;
+CREATE EXTENSION hll;
```
Added the lines

gurkanindibay commented:
[11/18/2021 1:42 PM] Marco Slot
well, that's if you want to allow anyone to connect from anywhere

[11/18/2021 1:42 PM] Marco Slot
to any database

[11/18/2021 1:42 PM] Marco Slot
we only want the citus user to connect to the citus database

[11/18/2021 1:42 PM] Marco Slot
so for that we could have: host citus citus all scram-sha-256

[11/18/2021 1:42 PM] Marco Slot
but there seems to be some background process (perhaps Citus?) that also connects as postgres user

[11/18/2021 1:43 PM] Marco Slot
which is being blocked by my pg_hba.conf

gurkanindibay commented:
> [11/18/2021 1:42 PM] Marco Slot well, that's if you want to allow anyone to connect from anywhere
> 
> [11/18/2021 1:42 PM] Marco Slot to any database
> 
> [11/18/2021 1:42 PM] Marco Slot we only want the citus user to connect to the citus database
> 
> [11/18/2021 1:42 PM] Marco Slot so for that we could have: host citus citus all scram-sha-256
> 
> [11/18/2021 1:42 PM] Marco Slot but there seems to be some background process (perhaps Citus?) that also connects as postgres user
> 
> [11/18/2021 1:43 PM] Marco Slot which is being blocked by my pg_hba.conf

[11/18/2021 1:42 PM] Marco Slot
well, that's if you want to allow anyone to connect from anywhere

[11/18/2021 1:42 PM] Marco Slot
to any database

[11/18/2021 1:42 PM] Marco Slot
we only want the citus user to connect to the citus database

[11/18/2021 1:42 PM] Marco Slot
so for that we could have: host citus citus all scram-sha-256

[11/18/2021 1:42 PM] Marco Slot
but there seems to be some background process (perhaps Citus?) that also connects as postgres user

[11/18/2021 1:43 PM] Marco Slot
which is being blocked by my pg_hba.conf

I performed some configurations that prevent postgres user to connect into the image

First I added connection policies to prevent postgres user to connect locally. 
https://github.com/citusdata/docker/blob/1fe2a6c149a397c697e362f24f55b7389c04b35c/hyperscale/pg_hba.conf#L94-L96
Then I added random possword to prevent connecting using host
https://github.com/citusdata/docker/blob/1fe2a6c149a397c697e362f24f55b7389c04b35c/hyperscale/docker-entrypoint.sh#L9

marcocitus reviewed (commented):
I seem to be getting an error now:
```
$ docker run --name test1 -p 5432:5432 -e CITUS_PASSWORD=Citus123 hyperscale_image
docker: Error response from daemon: OCI runtime create failed: container_linux.go:380: starting container process caused: exec: "docker-entrypoint.sh": executable file not found in $PATH: unknown.
```

marcocitus commented on hyperscale/Dockerfile:
```diff
@@ -0,0 +1,73 @@
+
+#ARG CITUS_PWD_COMMAND='\\tsed "s\/###CITUS_PASSWORD###\/${CITUS_PASSWORD}\/g" -i \/docker-entrypoint-initdb.d\/001-db-user-creation.sql\n'
+#RUN sed "/# if first arg looks like a flag.*/a ${CITUS_PWD_COMMAND}" -i /usr/local/bin/docker-entrypoint.sh
+#RUN sed "/file_env().*/i export POSTGRES_PASSWORD=$(echo $RANDOM | md5sum | head -c 20;)" -i /usr/local/bin/docker-entrypoint.sh
```
seems like we can remove these

gurkanindibay replied:
Fixed

marcocitus commented on hyperscale/Dockerfile:
```diff
@@ -0,0 +1,73 @@
+#RUN sed "/file_env().*/i export POSTGRES_PASSWORD=$(echo $RANDOM | md5sum | head -c 20;)" -i /usr/local/bin/docker-entrypoint.sh
+
+COPY hyperscale/docker-entrypoint.sh /usr/local/bin/docker-entrypoint.sh
+
```
I needed to add `RUN chmod 755 /usr/local/bin/docker-entrypoint.sh` here to make the image work.

gurkanindibay replied:
I was trying to replicate it in my local. You found the problem thanks. I added execute permission into the file in git repository. This will fix it. No need to add additional command on Dockerfile

marcocitus commented on hyperscale/docker-entrypoint.sh:
```diff
@@ -0,0 +1,354 @@
+#    ie: file_env 'XYZ_DB_PASSWORD' 'example'
+# (will allow for "$XYZ_DB_PASSWORD_FILE" to fill in the value of
+#  "$XYZ_DB_PASSWORD" from a file, especially for Docker's secrets feature)
+export POSTGRES_PASSWORD=$(echo $RANDOM | md5sum | head -c 20)
```
Maybe add a comment here to say that this is the main difference with the original and why we do this.

gurkanindibay replied:
Fixed

gurkanindibay commented:
> Probably a file permissions problem. Adding `RUN chmod 755 /usr/local/bin/docker-entrypoint.sh` seems to fix it.

I was trying to replicate it in my local. You found the problem thanks. I added execute permission into the file in git repository. This will fix it. No need to add additional command on Dockerfile

marcocitus commented on hyperscale/pg_hba.conf:
```diff
@@ -0,0 +1,101 @@
+# "local" is for Unix domain socket connections only
+# local   all             all                                     trust
+# IPv4 local connections:
+#host    all             all             127.0.0.1/32            trust
```
We need an extra rule here:
```
host citus citus 127.0.0.0/1 trust
```

Otherwise, certain operations (pg_cron, re-partition joins) fail.

gurkanindibay replied:
Fixed

hanefi commented on hyperscale/pg-13/Dockerfile:
```diff
@@ -0,0 +1,69 @@
+# This file is auto generated from it's template,
+# see citusdata/tools/packaging_automation/templates/docker/latest/latest.tmpl.dockerfile.
```
Is this comment correct? Is PG13 Dockerfile generated from latest.tmpl.dockerfile ?

I do not think that this comment is valid now.

Can we update all the comments and provide some explanations on how to update these dockerfiles ?

gurkanindibay replied:
Removed the template comments since these docker files are not template-generated

hanefi reviewed (approved):
It would be nice to have a section on the `README.md` with some details on how to use the docker images we create for hyperscale.

This may be done on a separate PR, and possibly we can write some documentation for the docker images with pg12, pg13, alpine etc.
